File: mvvmDatabinding/src/main/java/com/bigmikehoncho/mvvmdatabinding/ViewState.java
package com.bigmikehoncho.mvvmdatabinding;

import android.support.annotation.Nullable;

import io.reactivex.functions.Consumer;

public class ViewState<N extends Navigator> {
    private N navigator;
    private Consumer<N> nConsumer;

    public N getNavigator() {
        return navigator;
    }

    public void setNavigator(@Nullable N navigator) {
        this.navigator = navigator;

        tryRunning();
    }

    public void apply(@Nullable Consumer<N> nConsumer) {
        this.nConsumer = nConsumer;

        tryRunning();
    }

    private void tryRunning() {
        if (navigator != null && nConsumer != null) {
            try {
                this.nConsumer.accept(navigator);
                this.nConsumer = null;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }
}

File: mvvmDatabinding/src/main/java/com/bigmikehoncho/mvvmdatabinding/ViewState.kt
package com.bigmikehoncho.mvvmdatabinding

import io.reactivex.functions.Consumer
import java.lang.Exception

class ViewState<N : Navigator> {
    var navigator: N? = null
        set(navigator) {
            field = navigator

            tryRunning()
        }
    private var nConsumer: Consumer<N>? = null

    fun apply(nConsumer: Consumer<N>?) {
        this.nConsumer = nConsumer

        tryRunning()
    }

    private fun tryRunning() {
        if (this.navigator != null && nConsumer != null) {
            try {
                this.nConsumer!!.accept(this.navigator!!)
                this.nConsumer = null
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }
}
